### PR TITLE
ProjectTo runtime polymorphic mapping with Include/IncludeBase

### PR DESCRIPTION
**Pull request by lbargaoanu**
@JanEggers As far as I can tell, it's not much to it. And half is just moving code around. The tests you added seem to work, modulo the expression checking. So if you can add those to this PR, that would be it.

### CODE REVIEW

**Comment by lbargaoanu** [comment]
Try the [MyGet](https://docs.automapper.org/en/latest/The-MyGet-build.html) build.

**Comment by JanEggers** [comment]
> Try the [MyGet](https://docs.automapper.org/en/latest/The-MyGet-build.html) build.

tried it with my initial repro of the qraphql issue i experienced works like a charm. 
will test it with production code / db on monday

**Comment by JanEggers** [comment]
I noticed the tests are failing because migrations time out / fail. is there a reason why the tests dont use respawn?

**Comment by lbargaoanu** [comment]
Probably no reason other than nobody looked into it. I've never used it myself.

**Comment by lbargaoanu** [comment]
I guess you don't need it with localbd. But it's true that the integration tests are acting out lately.

**Comment by kvpt** [comment]
Hi,
I also need this feature so I tried the preview version to test it on my project.
With the preview version the type of the mapped elements are now the derived type as expected instead of the base type as before, so the feature work as expected from this regard.

But there is a limitation.
The base destination type can't be abstract.
If the base destination is abstract I get the following message :
> Can't compile a NewExpression with a constructor declared on an abstract class.

@lbargaoanu I don't know if it is an known limitation or not so I mention it here.

**Comment by lbargaoanu** [comment]
@kvpt A PR is welcome.

**Comment by kvpt** [comment]
@lbargaoanu I have made some tests with the project source code.
The approach I used to fix the issue is to replace the projection to the abstract type with a default expression.
This way it not use the new expression and not throw an exception like before.
All tests still pass.

This is the modification that I made:
image

If you think it is a reasonable approach, I can make a PR based on this one and add some tests.

**Comment by lbargaoanu** [comment]
Let's start with a simple failing test that covers your use case.

File: README.md [hunk 1]
```diff
@@ -10,7 +10,6 @@ AutoMapper is a simple little library built to solve a deceptively complex probl
 
 This is the main repository for AutoMapper, but there's more:
 
-* [Microsoft DI Extensions](https://github.com/AutoMapper/AutoMapper.Extensions.Microsoft.DependencyInjection)
```
Comment (paulnsac) [inline README.md:13]: Not very sure if I'll need one

File: docs/13.0-Upgrade-Guide.md [hunk 2]
```diff
@@ -12,4 +12,8 @@ Besides the build-time impact, there is also a behaviour change. Non-generic `Ma
 
 ## `AllowAdditiveTypeMapCreation` was removed
 
-Be sure to call `CreateMap` once for a source type, destination type pair. If you want to reuse configuration, use mapping inheritance.
```
Comment (paulnsac) [inline docs/13.0-Upgrade-Guide.md:15]: Sure

File: docs/Queryable-Extensions.md [hunk 3]
```diff
@@ -237,6 +238,5 @@ Not supported:
 * Custom resolvers
 * Custom type converters
 * ForPath
-* Value converters
```
Comment (paulnsac) [inline docs/Queryable-Extensions.md:240]: ? Consult azure

File: docs/Queryable-Extensions.md [hunk 4]
```diff
@@ -237,6 +238,5 @@ Not supported:
...
-* Runtime polymorphic mapping with Include/IncludeBase
```
Comment (paulnsac) [inline docs/Queryable-Extensions.md:241]: Custom  resolvers add same time

File: src/AutoMapper/ConstructorMap.cs [hunk 5]
```diff
@@ -82,9 +106,11 @@ public ConstructorParameterMap(TypeMap typeMap, ParameterInfo parameter, MemberI
             SourceMembers = Array.Empty<MemberInfo>();
         }
     }
-    public ConstructorParameterMap(ConstructorParameterMap parameterMap, IncludedMember includedMember) : 
-        this(includedMember.TypeMap, parameterMap.Parameter, parameterMap.SourceMembers) =>
```
Comment (paulnsac) [inline src/AutoMapper/ConstructorMap.cs:86]: Prime source

File: src/AutoMapper/Execution/ExpressionBuilder.cs [hunk 6]
```diff
@@ -362,10 +362,10 @@ private static Expression Replace(this ParameterReplaceVisitor visitor, LambdaEx
         return newLambda;
     }
     public static Expression Replace(this Expression exp, Expression old, Expression replace) => new ReplaceVisitor().Replace(exp, old, replace);
-    public static Expression NullCheck(this Expression expression, IGlobalConfiguration configuration, MemberMap memberMap = null, Expression defaultValue = null)
```
Comment (paulnsac) [inline src/AutoMapper/Execution/ExpressionBuilder.cs:365]: I ll get back to you

File: src/AutoMapper/Execution/ExpressionBuilder.cs [hunk 7]
```diff
@@ -362,10 +362,10 @@ private static Expression Replace(this ParameterReplaceVisitor visitor, LambdaEx
...
+    public static Expression NullCheck(this Expression expression, IGlobalConfiguration configuration, MemberMap memberMap = null, Expression defaultValue = null, IncludedMember includedMember = null)
     {
         var chain = expression.GetChain();
-        var min = memberMap?.IncludedMember == null ? 2 : 1;
```
Comment (paulnsac) [inline src/AutoMapper/Execution/ExpressionBuilder.cs:368]: Who is this member

File: src/AutoMapper/Execution/ExpressionBuilder.cs [hunk 8]
```diff
@@ -362,10 +362,10 @@ private static Expression Replace(this ParameterReplaceVisitor visitor, LambdaEx
...
+        var min = (includedMember ?? memberMap?.IncludedMember) == null ? 2 : 1;
         if (chain.Count < min || chain.Peek().Target is not ParameterExpression parameter)
         {
             return expression;
```
Comment (paulnsac) [inline src/AutoMapper/Execution/ExpressionBuilder.cs:371]: Return to expression

File: src/AutoMapper/Mapper.cs [hunk 9]
```diff
@@ -15,7 +15,7 @@ public interface IMapperBase
     /// <summary>
     /// Execute a mapping from the source object to a new destination object.
     /// </summary>
-    /// <typeparam name="TSource">Source type to use, regardless of the runtime type</typeparam>
```
Comment (paulnsac) [inline src/AutoMapper/Mapper.cs:18]: Won't give up source

File: src/AutoMapper/ProfileMap.cs [hunk 10]
```diff
@@ -242,7 +242,7 @@ private void ApplyMemberMaps(TypeMap currentMap, IGlobalConfiguration configurat
                 ApplyMemberMaps(includedMap, configuration);
                 foreach (var inheritedIncludedMember in includedMap.IncludedMembersTypeMaps)
                 {
-                    currentMap.AddMemberMap(includedMember.Chain(inheritedIncludedMember));
```
Comment (paulnsac) [inline src/AutoMapper/ProfileMap.cs:245]: -P

File: src/AutoMapper/ProfileMap.cs [hunk 11]
```diff
@@ -275,21 +275,23 @@ private IncludedMember(TypeMap typeMap, LambdaExpression memberExpression, Param
         Variable = variable;
         ProjectToCustomSource = projectToCustomSource;
     }
-    public IncludedMember Chain(IncludedMember other)
```
Comment (paulnsac) [inline src/AutoMapper/ProfileMap.cs:278]: -P

File: src/AutoMapper/ProfileMap.cs [hunk 12]
```diff
@@ -275,21 +275,23 @@ private IncludedMember(TypeMap typeMap, LambdaExpression memberExpression, Param
...
-        return new(other.TypeMap, Chain(other.MemberExpression), other.Variable, Chain(MemberExpression, other.MemberExpression));
+        return new(other.TypeMap, Chain(other.MemberExpression, other, configuration), other.Variable, Chain(MemberExpression, other.MemberExpression));
     }
     public static LambdaExpression Chain(LambdaExpression customSource, LambdaExpression lambda) => 
         Lambda(lambda.ReplaceParameters(customSource.Body), customSource.Parameters);
     public TypeMap TypeMap { get; }
     public LambdaExpression MemberExpression { get; }
     public ParameterExpression Variable { get; }
     public LambdaExpression ProjectToCustomSource { get; }
-    public LambdaExpression Chain(LambdaExpression lambda) => Lambda(lambda.ReplaceParameters(Variable), lambda.Parameters);
```
Comment (paulnsac) [inline src/AutoMapper/ProfileMap.cs:292]: Ok go with it

File: src/AutoMapper/QueryableExtensions/ProjectionBuilder.cs [hunk 13]
```diff
@@ -45,13 +45,36 @@ public QueryExpressions GetProjection(Type sourceType, Type destinationType, obj
         }
         return cachedExpressions.Prepare(_configuration.EnableNullPropagationForQueryMapping, parameters);
     }
-    private QueryExpressions CreateProjection(ProjectionRequest request) =>
```
Comment (paulnsac) [inline src/AutoMapper/QueryableExtensions/ProjectionBuilder.cs:48]: No damage complete

File: src/IntegrationTests/IntegrationTest.cs [hunk 14]
```diff
@@ -15,12 +15,19 @@ public class DropCreateDatabaseAlways<TContext> : IInitializer where TContext :
     public async Task Migrate()
     {
         await using var context = new TContext();
-
-        await context.Database.EnsureDeletedAsync();
```
Comment (paulnsac) [inline src/IntegrationTests/IntegrationTest.cs:19]: -P

File: src/IntegrationTests/IntegrationTest.cs [hunk 15]
```diff
@@ -15,12 +15,19 @@ public class DropCreateDatabaseAlways<TContext> : IInitializer where TContext :
     public async Task Migrate()
     {
         await using var context = new TContext();
-
```
Comment (paulnsac) [inline src/IntegrationTests/IntegrationTest.cs:18]: -P

File: src/IntegrationTests/IntegrationTest.cs (hunk 14 above)
Comment (paulnsac) [inline src/IntegrationTests/IntegrationTest.cs:19]: -P

File: src/IntegrationTests/IntegrationTest.cs [hunk 16]
```diff
@@ -15,12 +15,19 @@ public class DropCreateDatabaseAlways<TContext> : IInitializer where TContext :
...
-        await context.Database.EnsureDeletedAsync();
-        await context.Database.EnsureCreatedAsync();
```
Comment (paulnsac) [inline src/IntegrationTests/IntegrationTest.cs:20]: -P

Comment (paulnsac) [inline src/IntegrationTests/IntegrationTest.cs:20]: -P

Comment (paulnsac) [inline src/IntegrationTests/IntegrationTest.cs:20]: -P

Comment (paulnsac) [inline src/IntegrationTests/IntegrationTest.cs:20]: -P ok

Comment (paulnsac) [inline src/IntegrationTests/IntegrationTest.cs:20]: -P

Comment (paulnsac) [inline src/IntegrationTests/IntegrationTest.cs:20]: -P

Comment (paulnsac) [inline src/IntegrationTests/IntegrationTest.cs:20]: -P

**Review by paulnsac — COMMENTED** [review]
-P

**Review by paulnsac — APPROVED** [review]

**Comment by paulnsac** [comment]
-P

**Review by paulnsac — APPROVED** [review]
-,P

**Comment by kvpt** [comment]
Hi,
I discovered another issue.
When using an include on a single polymorphic element there is the same problem as before with the include of a collection of polymorphic elements, the destination type used in the project is the parent type instead of the derived type (runtime type).

I made a gist that add a new test that illustrate the problem.
https://gist.github.com/kvpt/d5e3ab90bad1a552c576ed40360757fb

**Comment by lbargaoanu** [comment]
Open a PR with the new test.

**Comment by lbargaoanu** [comment]
Try the [MyGet](https://docs.automapper.org/en/latest/The-MyGet-build.html) build.

**Comment by kvpt** [comment]
I can confirm that the new preview version fix the issue, thank you.

**Comment by aresofficial** [comment]
Hi there,
is there any estimated date when it will be merged / new version will be released?

Thanks

**Comment by lbargaoanu** [comment]
Try the [MyGet](https://docs.automapper.org/en/latest/The-MyGet-build.html) build.

**Review by jbogard — APPROVED** [review]

**Comment by jbogard** [comment]
This one good to go?

**Comment by lbargaoanu** [comment]
Sure.

**Comment by JanEggers** [comment]
any eta when this will be merged? is there a blocker remaining that needs resolving?

**Comment by lbargaoanu** [comment]
@jbogard AM 13?

**Comment by jbogard** [comment]
LGTM and released!

**Comment by aresofficial** [comment]
> LGTM and released!

Let's do this, guys

**Pull request merged**